### DISCUSSION TITLE
GotoX在OpenWrt上成功运行

### DISCUSSION THREAD

**Discussion by Peytonlan** [SeaHOH/GotoX#58] [展示]
GotoX是头两天才注意到, 因支持sniproxy, 尝试用了下, 还不错, 感谢 @SeaHOH

貌似现在的风潮都在往路由器上挤, 当然GotoX也不能例外
方法比较通用, 只要软件对内存要求别太过份, 一般采用这种方式都可行
但对操作者的技能有些要求, 看不明白的可以忽略, 毕竟刷机有风险
看得明白的, 简述下就能理解, 因此以下不涉及具体细节

router: TL-WDR4320

host OS: LEDE 17.01 "Reboot" 自编译
  要支持LXC, 必须要自己编译系统
  Debian的mips可执行文件是按硬浮点编译的, 所以 CONFIG_MIPS_FPU_EMULATOR=y 一定要设置, 不然无法在路由器上执行

guest OS: Debian stretch
  官方的 LXC image 源是不包含mips架构的, 所以需要自己制作或利用第三方现成的
  GotoX比较新潮, dnslib & brotlipy 在stretch上需要通过pip安装
  路由器性能强劲的, 可以用现成的rootfs http://downloads.debwrt.net/rootfs/snapshots/mips/ 试试
  路由器性能弱的, 直接在路由器上通过pip安装会比较痛苦, 而且很可能成不了
  推荐用QEMU自己安装mips版的Debian stretch, 自己制作rootfs, 好处是比路由器快, 软件依赖都可以在虚拟机里预先配置好
  GotoX在mips虚拟机里安装配置好, 是可以成功运行的

lxc-info -n gotox
Name:           gotox
State:          RUNNING
CPU use:        6080.48 seconds
Memory use:     20.91 MiB
KMem use:       0 bytes
Link:           gotox
 TX bytes:      3.35 GiB
 RX bytes:      3.53 GiB
 Total bytes:   6.88 GiB

uptime
  up 12:31,  load average: 0.57, 0.52, 0.50

内存, cpu, 负载都不高, 挺好的

**Comment by SeaHOH** [comment]
玩的高兴 :hand: 

不过，这么用的话，需要给自动规则搞个网页 UI 来配置，不然不方便修改。
需要吗？要就搞一个。

**Comment by Peytonlan** [comment]
谢谢 :raised_hand:

我是不需要这个功能的, 也许会有别的人想要
应该不用急吧, 看看其他人的反馈再定

**Comment by SeaHOH** [comment]
之前用得人可能多些，现在估计不超 10 个 ，那就保持现状吧。

**Comment by SeaHOH** [comment]
设置基本上跟谷歌 IP 差不多
```ini
[gae]
iplist = sni_all # 使用自己新建的列表
servername = # 关键设置，不会设置就留空，默认值应该能正常工作
[iplist]
sni_all = ip1|ip2|... # SNIProxy IP，如果是国内的，需要自行查找可用的 servername
```

**Comment by SeaHOH** [comment]
>这个是什么用途？

给 GAE 使用前置代理时用的 IP 列表
除非有特别需求一般不用设置 `iplist2p` 或修改 `google_2p` 列表

>appid支持域名绑定吗？

没有

**Comment by SeaHOH** [comment]
>难道已经生效了？怎么理解.

启用 GAE 前置，这个列表才有意义。
解析后通常是台湾 IP，这是能正常工作的。
你也可以根据你使用的前置的地理位置修改这个列表，以获取更好的链接性能。

**Comment by Peytonlan** [comment]
@SeaHOH
在没有查看代码的情况下, 我来猜测下 google_2p 的意图
假定使用的 GAE 前置代理的位置在日本, google_2p 设置成 www.google.co.jp 有可能获得更好的性能, 因为www.google.co.jp 解析到的ip可能会和代理的距离最近
google_2p 设置的域名实际上就是, 当透过 GAE 前置代理访问 GAE 服务时, 所用的 servername
这样理解, 对不对

**Comment by SeaHOH** [comment]
@Peytonlan 
可以直接填 IP，servername 有单独的设置项，google_2p 只是个 IP 列表而已，单独列出只是为了方便设定默认值。

**Comment by Peytonlan** [comment]
谢谢澄清

**Comment by wpyok500** [comment]
你们还有tls  ip可用？？？还是用是sni?

**Comment by SeaHOH** [comment]
反正我用的是 SNIProxy，没有也懒得扫，不知道有没有可用的 IP。
其实国内有部分 IP（非谷歌官方）可用，不知道到底是怎么回事。

**Comment by Peytonlan** [comment]
sni

**Comment by SeaHOH** [comment]
学校啊，事到如今也不想吐槽了。
